### ISSUE TITLE
Getting False Positive REQUEST-949-BLOCKING-EVALUATION

### ISSUE THREAD

**Issue by DeepMalh44** [SpiderLabs/owasp-modsecurity-crs#1449]
Hi All,
We enabled WAF on PROD environment and on the web-application we are getting some issues. Basically as follows:-

PROD2-APPGW | OWASP | 3 | Default | 0 | Detected | Global | Warning.   Operator GE matched 5 at TX:anomaly_score. | rules/REQUEST-949-BLOCKING-EVALUATION.conf | 57 | quote.threeinsurance.com | Mandatory rule. Cannot be disabled. Inbound   Anomaly Score Exceeded (Total Score: 5) | /Content/styles/images/close-small.svg
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --

We have checked with Application owners about it and they say this is a valid request and no issue, can someone help me with why we getting this false positive if there is anything we can do to avoid it or ask App owners to change?

I am trying to look for a place whcih can explain me these rules plain and simple which I forward to the business owners as well.

Thank you,
Ketan

**Comment by dune73** [comment]
Sorry for the inconvenience @ketaanhshah. Unfortunately, your report only indicates _that_ a rule has triggered, but not which one. (In fact you report the evaluation rule, which evaluates the result of a different rule that detected an attack; and that rule is the culprit here).

You may want to check out my CRS tuning tutorial at https://netnea.com/apache-tutorials. It explains how to get rid of false positives as these.

If you have a false positives in the default installation and you can provide us with the payload that triggers the alert, we'll be happy to look into it. If not, we're going to close this issue in a few days.

**Comment by DeepMalh44** [comment]
Thank you @dune73 Christian, I am not the person in my team with WAF and security background but I will do the most to find out what payload is generating these errors. I have attached the log file that I got from Infra team about one of the applications which was throwing the errors. I have sent this log to Application team asking them for individual request to clarify if all are false positives. But I see even for some fonts and images and stylesheet files there is a trigger on WAF rule.


[errs - Copy.xlsx](https://github.com/SpiderLabs/owasp-modsecurity-crs/files/3276772/errs.-.Copy.xlsx)

**Comment by DeepMalh44** [comment]
@dune73 My Application team actually got back to me with some reasoning on why there could be the issues, for the last one I agree its a legitimate request but others are something we probably need help on fixing from our CRS Rule end, any advice?:-


Count | Rule Group | Comments | Action
-- | -- | -- | --
17 | 942-APPLICATION-ATTACK-SQLI | This is being raised because of the request verification   token.  This token is a random 92 character string that sometimes can   contain – and is when the rule is triggered. | None
3 | This is being raised because a # or ‘ is being included   within the address lookup function.  This is likely due to the user   trying to input this into the auto complete function. | None
6 | The _section query string value is wrapped with an   apostrophe. | TBC – Shim is investigating if we can remove the   apostrophe.
31 | Default | This is not clear on the route cause, some of the errors   are because some query string values have been added incorrectly but from   where and why is not possible to determine from these logs.  Others are   from non-transactional pages e.g. /session/quotesessionexpired which again   are not clear from the logs. | More data required
4 | 920-PROTOCOL-ENFORCEMENT | Appears a legitimate block – the User-Agent header was   missing from the request. | None

**Comment by DeepMalh44** [comment]
@dune73 Christian any help on the above, do you think we will have to write our own custom rules and these custom rules will superceed the out of the box CRS 3.0 rules so even if they are kept enabled it wont hurt as custom rule will allow the requests even though ootb rules will block them? Or u think we have to disable the ootb rules? Also any guidance on how we can write the custom rules for above errors?

**Comment by csanders-git** [comment]
Hey @ketaanhshah  in order to answer that we'll need to know the ID of the rule(s) that fired and the content that triggered that rule. Both of these would be included in the error.log (by default) on the system where the rule triggered. In general it is possible to create exceptions for false positive or in some cases disable the rule, if they cannot be avoided (typically this is due to legit but abnormal request parameters (i.e sending HTML via a parameter).

**Comment by DeepMalh44** [comment]
Thank you @csanders-git Chaim, I will try to get more details from the log and send it. Meanwhile can you let me know if we write a custom rule and keep the out of the box rule enabled, and in our custom rule we enable/allow the request which is blocked in ootb rule in that case will the custom rule superceed and allow OR we have to disable the ootb BLOCK rule and let custom rule take care of it.

Basically I am looking for a way whether I need to disable a rule completely or I can keep it on and just write a override custom rule?

**Comment by DeepMalh44** [comment]
@csanders-git Chaims here is the following table with Rule ID


App | Rule ID | Rule Group | Message | Comment
-- | -- | -- | -- | --
CMS | 942260 |   | Detects   basic SQL authentication bypass attempts 2/3 | High   chances of false-positive. All the requests looks genuine asking for media,   fonts, css, scripts, favicon, dclid, autodiscover
CMS | 942340 |   | Detects   basic SQL authentication bypass attempts 3/3
CMS | 942330 |   | Detects   classic SQL injection probings 1/2
CMS | 942370 |   | Detects   classic SQL injection probings 2/2
CMS | 942200 |   | Detects   MySQL comment-/space-obfuscated injections and backtick termination
CMS | 942300 |   | Detects   MySQL comments, conditions and ch(a)r injections
CMS | 942440 |   | SQL   Comment Sequence Detected.

**Comment by ketaanhshah** [comment]
Some more additional details here:-

Resource | ruleSetType_s | ruleSetVersion_s | ruleGroup_s | ruleId_s | action_s | site_s | details_message_s | details_data_s | details_file_s | details_line_s | hostname_s | Message | requestUri_s | count_
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
PROD2-APPGW | OWASP | 3 | 942-APPLICATION-ATTACK-SQLI | 942370 | Matched | Global | Warning. Pattern match   "(?i:(?:[\"'`]\\s*?\\*.+(?:x?or\|div\|like\|between\|and\|id)\\W*?[\"'`]\\d)\|(?:\\^[\"'`])\|(?:^[\\w\\s\"'`-]+(?<=and\\s)(?<=or\|xor\|div\|like\|between\|and\\s)(?<=xor\\s)(?<=nand\\s)(?<=not\\s)(?<=\\\|\\\|)(?<=\\&\\&)\\w+\\()\|(?:[\"'`][\\s\\d]*?[^\\w\\s]+\\W*?\\d\   ..." at REQUEST_COOKIES:fs_uid. | Matched Data: `4703519480414208:62 found within REQUEST_COOKIES:fs_uid:   rs.fullstory.com`M4G4J`4703519480414208:6216073757261824 | rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf | 841 |   | Detects classic SQL injection probings 2/2 | / | 1
PROD2-APPGW | OWASP | 3 | Default | 0 | Detected | Global | Warning. Operator GE matched 5 at   TX:anomaly_score. | rules/REQUEST-949-BLOCKING-EVALUATION.conf | 57 |   | Mandatory rule. Cannot be disabled. Inbound Anomaly Score Exceeded (Total   Score: 5) | / | 14
PROD2-APPGW | OWASP | 3 | Default | 0 | Detected | Global | Warning. Operator GE matched 5 at   TX:inbound_anomaly_score. | rules/RESPONSE-980-CORRELATION.conf | 73 |   | Mandatory rule. Cannot be disabled. Inbound Anomaly Score Exceeded (Total   Inbound Score: 5 - SQLI=5,XSS=0,RFI=0,LFI=0,RCE=0,PHPI=0,HTTP=0,SESS=0):   Detects classic SQL injection probings 2/2 | / | 13
PROD2-APPGW | OWASP | 3 | 942-APPLICATION-ATTACK-SQLI | 942370 | Matched | Global | Warning. Pattern match   "(?i:(?:[\"'`]\\s*?\\*.+(?:x?or\|div\|like\|between\|and\|id)\\W*?[\"'`]\\d)\|(?:\\^[\"'`])\|(?:^[\\w\\s\"'`-]+(?<=and\\s)(?<=or\|xor\|div\|like\|between\|and\\s)(?<=xor\\s)(?<=nand\\s)(?<=not\\s)(?<=\\\|\\\|)(?<=\\&\\&)\\w+\\()\|(?:[\"'`][\\s\\d]*?[^\\w\\s]+\\W*?\\d\   ..." at REQUEST_COOKIES:fs_uid. | Matched Data: `4703519480414208:62 found within REQUEST_COOKIES:fs_uid:   rs.fullstory.com`M4G4J`4703519480414208:6216073757261824 | rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf | 841 |   | Detects classic SQL injection probings 2/2 | /media/1681/jqntgio3p0r-z-0-y-paper-stack.mp4 | 3
PROD2-APPGW | OWASP | 3 | Default | 0 | Detected | Global | Warning. Operator GE matched 5 at   TX:anomaly_score. | rules/REQUEST-949-BLOCKING-EVALUATION.conf | 57 |   | Mandatory rule. Cannot be disabled. Inbound Anomaly Score Exceeded (Total   Score: 5) | /media/1681/jqntgio3p0r-z-0-y-paper-stack.mp4 | 80
PROD2-APPGW | OWASP | 3 | Default | 0 | Detected | Global | Warning. Operator GE matched 5 at   TX:inbound_anomaly_score. | rules/RESPONSE-980-CORRELATION.conf | 73 |   | Mandatory rule. Cannot be disabled. Inbound Anomaly Score Exceeded (Total   Inbound Score: 5 - SQLI=5,XSS=0,RFI=0,LFI=0,RCE=0,PHPI=0,HTTP=0,SESS=0):   Detects classic SQL injection probings 2/2 | /media/1681/jqntgio3p0r-z-0-y-paper-stack.mp4 | 80
PROD2-APPGW | OWASP | 3 | 942-APPLICATION-ATTACK-SQLI | 942370 | Matched | Global | Warning. Pattern match   "(?i:(?:[\"'`]\\s*?\\*.+(?:x?or\|div\|like\|between\|and\|id)\\W*?[\"'`]\\d)\|(?:\\^[\"'`])\|(?:^[\\w\\s\"'`-]+(?<=and\\s)(?<=or\|xor\|div\|like\|between\|and\\s)(?<=xor\\s)(?<=nand\\s)(?<=not\\s)(?<=\\\|\\\|)(?<=\\&\\&)\\w+\\()\|(?:[\"'`][\\s\\d]*?[^\\w\\s]+\\W*?\\d\   ..." at REQUEST_COOKIES:fs_uid. | Matched Data: `6644820941668352:48 found within REQUEST_COOKIES:fs_uid:   rs.fullstory.com`M4G4J`6644820941668352:4856609099677696 | rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf | 841 |   | Detects classic SQL injection probings 2/2 | /the-policy/ | 1
PROD2-APPGW | OWASP | 3 | Default | 0 | Detected | Global | Warning. Operator GE matched 5 at   TX:anomaly_score. | rules/REQUEST-949-BLOCKING-EVALUATION.conf | 57 |   | Mandatory rule. Cannot be disabled. Inbound Anomaly Score Exceeded (Total   Score: 5) | /the-policy/ | 5
PROD2-APPGW | OWASP | 3 | 942-APPLICATION-ATTACK-SQLI | 942370 | Matched | Global | Warning. Pattern match   "(?i:(?:[\"'`]\\s*?\\*.+(?:x?or\|div\|like\|between\|and\|id)\\W*?[\"'`]\\d)\|(?:\\^[\"'`])\|(?:^[\\w\\s\"'`-]+(?<=and\\s)(?<=or\|xor\|div\|like\|between\|and\\s)(?<=xor\\s)(?<=nand\\s)(?<=not\\s)(?<=\\\|\\\|)(?<=\\&\\&)\\w+\\()\|(?:[\"'`][\\s\\d]*?[^\\w\\s]+\\W*?\\d\   ..." at REQUEST_COOKIES:fs_uid. | Matched Data: `5179329043693568:46 found within REQUEST_COOKIES:fs_uid:   rs.fullstory.com`M4G4J`5179329043693568:4650395164147712 | rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf | 841 |   | Detects classic SQL injection probings 2/2 | /media/1439/policy-peek-2x.png | 1
PROD2-APPGW | OWASP | 3 | Default | 0 | Detected | Global | Warning. Operator GE matched 5 at   TX:anomaly_score. | rules/REQUEST-949-BLOCKING-EVALUATION.conf | 57 |   | Mandatory rule. Cannot be disabled. Inbound Anomaly Score Exceeded (Total   Score: 5) | /media/1439/policy-peek-2x.png | 23
PROD2-APPGW | OWASP | 3 | Default | 0 | Detected | Global | Warning. Operator GE matched 5 at   TX:inbound_anomaly_score. | rules/RESPONSE-980-CORRELATION.conf | 73 |   | Mandatory rule. Cannot be disabled. Inbound Anomaly Score Exceeded (Total   Inbound Score: 5 - SQLI=5,XSS=0,RFI=0,LFI=0,RCE=0,PHPI=0,HTTP=0,SESS=0):   Detects classic SQL injection probings 2/2 | /media/1439/policy-peek-2x.png | 23
PROD2-APPGW | OWASP | 3 | 942-APPLICATION-ATTACK-SQLI | 942370 | Matched | Global | Warning. Pattern match   "(?i:(?:[\"'`]\\s*?\\*.+(?:x?or\|div\|like\|between\|and\|id)\\W*?[\"'`]\\d)\|(?:\\^[\"'`])\|(?:^[\\w\\s\"'`-]+(?<=and\\s)(?<=or\|xor\|div\|like\|between\|and\\s)(?<=xor\\s)(?<=nand\\s)(?<=not\\s)(?<=\\\|\\\|)(?<=\\&\\&)\\w+\\()\|(?:[\"'`][\\s\\d]*?[^\\w\\s]+\\W*?\\d\   ..." at REQUEST_COOKIES:fs_uid. | Matched Data: `5179329043693568:46 found within REQUEST_COOKIES:fs_uid:   rs.fullstory.com`M4G4J`5179329043693568:4650395164147712 | rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf | 841 |   | Detects classic SQL injection probings 2/2 | /media/1268/propertyassets-3x.png | 1
PROD2-APPGW | OWASP | 3 | Default | 0 | Detected | Global | Warning. Operator GE matched 5 at   TX:anomaly_score. | rules/REQUEST-949-BLOCKING-EVALUATION.conf | 57 |   | Mandatory rule. Cannot be disabled. Inbound Anomaly Score Exceeded (Total   Score: 5) | /media/1268/propertyassets-3x.png | 25
PROD2-APPGW | OWASP | 3 | Default | 0 | Detected | Global | Warning. Operator GE matched 5 at   TX:inbound_anomaly_score. | rules/RESPONSE-980-CORRELATION.conf | 73 |   | Mandatory rule. Cannot be disabled. Inbound Anomaly Score Exceeded (Total   Inbound Score: 5 - SQLI=5,XSS=0,RFI=0,LFI=0,RCE=0,PHPI=0,HTTP=0,SESS=0):   Detects classic SQL injection probings 2/2 | /media/1268/propertyassets-3x.png | 25
PROD2-APPGW | OWASP | 3 | 942-APPLICATION-ATTACK-SQLI | 942370 | Matched | Global | Warning. Pattern match   "(?i:(?:[\"'`]\\s*?\\*.+(?:x?or\|div\|like\|between\|and\|id)\\W*?[\"'`]\\d)\|(?:\\^[\"'`])\|(?:^[\\w\\s\"'`-]+(?<=and\\s)(?<=or\|xor\|div\|like\|between\|and\\s)(?<=xor\\s)(?<=nand\\s)(?<=not\\s)(?<=\\\|\\\|)(?<=\\&\\&)\\w+\\()\|(?:[\"'`][\\s\\d]*?[^\\w\\s]+\\W*?\\d\   ..." at REQUEST_COOKIES:fs_uid. | Matched Data: `5179329043693568:46 found within REQUEST_COOKIES:fs_uid:   rs.fullstory.com`M4G4J`5179329043693568:4650395164147712 | rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf | 841 |   | Detects classic SQL injection probings 2/2 | /media/1514/img-policy-operations.jpg | 1

**Comment by DeepMalh44** [comment]
Resource | ruleSetType_s | ruleSetVersion_s | ruleGroup_s | ruleId_s | action_s | site_s | details_message_s | details_data_s | details_file_s | details_line_s | hostname_s | Message | requestUri_s | count_
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
PROD2-APPGW | OWASP | 3 | 942-APPLICATION-ATTACK-SQLI | 942370 | Matched | Global | Warning. Pattern match   "(?i:(?:[\"'`]\\s*?\\*.+(?:x?or\|div\|like\|between\|and\|id)\\W*?[\"'`]\\d)\|(?:\\^[\"'`])\|(?:^[\\w\\s\"'`-]+(?<=and\\s)(?<=or\|xor\|div\|like\|between\|and\\s)(?<=xor\\s)(?<=nand\\s)(?<=not\\s)(?<=\\\|\\\|)(?<=\\&\\&)\\w+\\()\|(?:[\"'`][\\s\\d]*?[^\\w\\s]+\\W*?\\d\   ..." at REQUEST_COOKIES:fs_uid. | Matched Data: `4703519480414208:62 found within REQUEST_COOKIES:fs_uid:   rs.fullstory.com`M4G4J`4703519480414208:6216073757261824 | rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf | 841 |   | Detects classic SQL injection probings 2/2 | / | 1
PROD2-APPGW | OWASP | 3 | Default | 0 | Detected | Global | Warning. Operator GE matched 5 at   TX:anomaly_score. | rules/REQUEST-949-BLOCKING-EVALUATION.conf | 57 |   | Mandatory rule. Cannot be disabled. Inbound Anomaly Score Exceeded (Total   Score: 5) | / | 14
PROD2-APPGW | OWASP | 3 | Default | 0 | Detected | Global | Warning. Operator GE matched 5 at   TX:inbound_anomaly_score. | rules/RESPONSE-980-CORRELATION.conf | 73 |   | Mandatory rule. Cannot be disabled. Inbound Anomaly Score Exceeded (Total   Inbound Score: 5 - SQLI=5,XSS=0,RFI=0,LFI=0,RCE=0,PHPI=0,HTTP=0,SESS=0):   Detects classic SQL injection probings 2/2 | / | 13

**Comment by csanders-git** [comment]
Looks like all of those  FPs are at paranoia level 2 -- this configuration level is expected to have some false positives. my recommendation would be to start in PL1.
You can then use `setvar:tx.executing_paranoia_level=2` in the crs-setup.conf to only alert (but not contribute to the anaomly score for anything at PL2) for anything in a higher paranoia level. You can then build a list of exclusions based on false positives over time before re-enabling this mode. 

Does that make sense?

**Comment by DeepMalh44** [comment]
Thank you @csanders-git , I am so new to all this that first I need to figure out how I can modify the Paranoia levels in Azure to level=2 using the above command you mentioned to modify the crs-setup.conf. Can you please help me with some link to documentation on how to set it so that it takes effect on Azure environment we are using. We have multiple env. DEV, UAT and PROD.. so we can start with DEV first.

**Comment by csanders-git** [comment]
I have no idea how azure CRS works or how it is configured in their environment :-\ -- perhaps someone else may be able to help you who has used it. The docs on setting this value are in line, you can see them here on git by going here: https://github.com/SpiderLabs/owasp-modsecurity-crs/blob/v3.2/dev/crs-setup.conf.example#L131.

**Comment by DeepMalh44** [comment]
Thank you @csanders-git Chaim appreciate your quick help, I only wish now if I can get some answer on how we can change these levels on Azure as we have WAF enabled on Azure and our applications are hosted on a IAAS VM on Azure.

**Comment by csanders-git** [comment]
let us know if/when you figure it out.

**Comment by DeepMalh44** [comment]
Thank you @csanders-git Microsoft Engineer just confirmed that it is not possible at this moment to change it. Following is the github issue # for it: https://github.com/MicrosoftDocs/azure-docs/issues/33205

**Comment by csanders-git** [comment]
wow, surprising, however,  looking at the docs -- it seems you are probably able to quickly solve your problem. They support REQUEST_COOKIES, and all your problems documented seem to stem from one cookie, REQUEST_COOKIES:fs_uid -- looks like you could (at the cost of some security) get this to work as expected.

**Comment by csanders-git** [comment]
I'm going to close this issue though, let me know if you find a more satisfying solution.

**Comment by DeepMalh44** [comment]
@csanders-git Thank you sir, yes we used exclusion rule to exclude fs_uid and go ahead!
